lib/ancestor_builder.c

```diff
@@ -181,23 +202,62 @@ ancestor_builder_print_state(ancestor_builder_t *self, FILE *out)
     return 0;
 }
 
-int
-ancestor_builder_alloc(
-    ancestor_builder_t *self, size_t num_samples, size_t max_sites, int flags)
+#ifdef MMAP_GENOTYPES
+
+static int
+ancestor_builder_make_genotype_mmap(ancestor_builder_t *self)
 {
+
     int ret = 0;
-    unsigned long max_size = 1024 * 1024;
 
-    memset(self, 0, sizeof(ancestor_builder_t));
-    if (num_samples <= 1) {
-        ret = TSI_ERR_BAD_NUM_SAMPLES;
+    self->mmap_size = self->max_sites * self->encoded_genotypes_size;
+    if (ftruncate(self->mmap_fd, (off_t) self->mmap_size) != 0) {
+        ret = TSI_ERR_IO;
         goto out;
     }
+    self->mmap_buffer = mmap(
+        NULL, self->mmap_size, PROT_READ | PROT_WRITE, MAP_SHARED, self->mmap_fd, 0);
```

Comment:
This is probably unnecessary complexity, but once the genotypes are loaded into the mmap, we could speed things up by re-opening it with only `PROT_READ`?

Reply:
Good idea. I think the benefit would be marginal, and it might cause a full flush of the file.

Worth following up if you like though, we can open an issue?